File: Материалы для проверки решений рег этап 2015/Пробный тур/Задача Z Конфеты/sweets_sk_070.py
#
# Author: Sergey Kopeliovich (Burunduk30@gmail.com)
# Date: 2013.11.15
#

import sys

NAME = "sweets"

inf = open(NAME + ".in", "r")
out = open(NAME + ".out", "w")

n, a, b, c = list(map(int, inf.readline().split()))

best = 0
rx, ry, rz = 0, 0, 0
x0 = n // (3 * a)
y0 = n // (3 * b)
mx = n // a
M = 3000

for x in range(max(1, x0 - M), min(mx, x0 + M) + 1):
  my = (n - x * a) // b
  for y in range(max(1, y0 - M), min(my, y0 + M) + 1):
    z = (n - a * x - b * y) // c
    tmp = x * y * z
    if (tmp - best > 0):
      best, rx, ry, rz = tmp, x, y, z

print(rx * a, ry * b, rz * c, file=out)

inf.close()
out.close()

File: Материалы для проверки решений рег этап 2015/Пробный тур/Задача Z Конфеты/sweets_sk_070.cpp
/**
 * Author: Sergey Kopeliovich (Burunduk30@gmail.com)
 * Date: 2013.11.15
 */

#include <cstdio>
#include <cmath>
#include <cassert>
#include <iostream>
#include <algorithm>

using namespace std;

#define forn(i, n) for (int i = 0; i < (int)(n); i++)

typedef long double dbl;

int main()
{
  #define NAME "sweets"
  assert(freopen(NAME ".in", "r", stdin));
  assert(freopen(NAME ".out", "w", stdout));

  int n, a, b, c;
  assert(scanf("%d%d%d%d", &n, &a, &b, &c) == 4);

  dbl best = 0;
  int rx = 0, ry = 0, rz = 0;
  int x0 = n / (3 * a);
  int y0 = n / (3 * b);

  int cnt = (int)1e7; // numer of operations
  int MX = sqrt(cnt) / 2, mx = n / a;
  int len = min(mx, x0 + MX) - max(1, x0 - MX) + 1;
  int MY = cnt / max(len, 1) / 2;
  for (int x = max(1, x0 - MX); x <= min(mx, x0 + MX); x++)
  {
    int my = (n - x * a) / b;
    for (int y = max(1, y0 - MY); y <= min(my, y0 + MY); y++)
    {
      int z = (n - a * x - b * y) / c;
      dbl tmp = (dbl)x * y * z;
      if (!best || tmp - best > 0)
        best = tmp, rx = x, ry = y, rz = z;
    }
  }
  
  printf("%d %d %d\n", rx * a, ry * b, rz * c);
  return 0;
}
